File: app/src/main/java/tipz/viola/search/SuggestionAdapter.java
/*
 * Copyright (C) 2022-2023 Tipz Team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package tipz.viola.search;

import android.content.Context;
import android.graphics.Typeface;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.style.StyleSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Filter;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.appcompat.widget.AppCompatTextView;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import tipz.viola.R;
import tipz.viola.webview.VioWebViewActivity;

/*
    "Inspired" by LineageOS' Jelly
 */

public class SuggestionAdapter extends ArrayAdapter<String> {
    private final ArrayList<String> mItems = new ArrayList<>();
    private final Context mContext;
    private final LayoutInflater mInflater;
    private final ItemFilter mFilter;
    private String mQueryText;

    public SuggestionAdapter(@NonNull Context context, int resource) {
        super(context, resource);
        mContext = context;
        mInflater = LayoutInflater.from(mContext);
        mFilter = new ItemFilter();
    }

    @Override
    public int getCount() {
        return mItems.size();
    }

    @Override
    public String getItem(int position) {
        return mItems.get(position);
    }

    @Override
    public long getItemId(int position) {
        return 0;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        if (convertView == null)
            convertView = mInflater.inflate(R.layout.list_item_suggestions, parent, false);

        AppCompatTextView title = convertView.findViewById(android.R.id.text1);
        AppCompatImageView copy_to_search_bar_button = convertView.findViewById(R.id.copy_to_search_bar_button);
        String suggestion = mItems.get(position);

        if (mQueryText != null) {
            SpannableStringBuilder spannable = new SpannableStringBuilder(suggestion);
            String lcSuggestion = suggestion.toLowerCase(Locale.getDefault());
            int queryTextPos = lcSuggestion.indexOf(mQueryText);
            while (queryTextPos >= 0) {
                spannable.setSpan(new StyleSpan(Typeface.BOLD),
                        queryTextPos, queryTextPos + mQueryText.length(),
                        Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                queryTextPos = lcSuggestion.indexOf(mQueryText, queryTextPos + mQueryText.length());
            }
            title.setText(spannable);
        } else {
            title.setText(suggestion);
        }

        copy_to_search_bar_button.setOnClickListener(view -> ((VioWebViewActivity) mContext).onUrlUpdated(suggestion, suggestion.length()));
        return convertView;
    }

    @Override
    public Filter getFilter() {
        return mFilter;
    }

    private class ItemFilter extends Filter {
        @Override
        protected FilterResults performFiltering(@Nullable CharSequence constraint) {
            FilterResults results = new FilterResults();
            if (constraint == null || constraint.length() == 0)
                return results;

            SuggestionProvider provider = getProvider();
            String query = constraint.toString().toLowerCase(Locale.getDefault()).trim();

            List<String> items = provider.fetchResults(query);
            results.count = items.size();
            results.values = items;

            return results;
        }

        @Override
        @SuppressWarnings("unchecked")
        protected void publishResults(@Nullable CharSequence constraint, FilterResults results) {
            mItems.clear();
            if (results.values != null) {
                mItems.addAll((List<String>) results.values);
                mQueryText = constraint != null
                        ? constraint.toString().toLowerCase(Locale.getDefault()).trim() : null;
            }
            notifyDataSetChanged();
        }
    }

    private SuggestionProvider getProvider() {
        return new SuggestionProvider(mContext);
    }
}

File: app/src/main/java/tipz/viola/search/SuggestionAdapter.kt
// Copyright (c) 2020-2025 The LineageOS Project
// Copyright (c) 2022-2026 Tipz Team
// SPDX-License-Identifier: Apache-2.0

package tipz.viola.search

import android.graphics.Typeface
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.style.StyleSpan
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.Filter
import android.widget.Filterable
import android.widget.Space
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.view.isVisible
import com.google.android.material.behavior.SwipeDismissBehavior
import com.google.android.material.snackbar.BaseTransientBottomBar
import com.google.android.material.snackbar.Snackbar
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import tipz.viola.R
import tipz.viola.databinding.TemplateTextSuggestionsBinding
import tipz.viola.ext.copyClipboard
import tipz.viola.ext.getOnSurfaceColor
import tipz.viola.ext.setStartAligned
import tipz.viola.settings.SettingsKeys
import tipz.viola.utils.UrlUtils
import tipz.viola.webview.VWebViewActivity
import java.util.Locale

class SuggestionAdapter(private val context: VWebViewActivity) : BaseAdapter(), Filterable {
    private var items = listOf<SuggestItem>()
    private val filter = ItemFilter()
    private var queryText: String? = null

    var enableFiltering = true
        set(value) {
            field = value
            queryText = ""
            items = listOf()
        }

    override fun getCount() = items.size

    override fun getItem(position: Int) = items[position]

    override fun getItemId(position: Int) = 0L

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val binding: TemplateTextSuggestionsBinding
        var itemView = convertView

        if (convertView == null || convertView !is ConstraintLayout) {
            binding = TemplateTextSuggestionsBinding.inflate(
                LayoutInflater.from(context), parent, false)
            itemView = binding.root
        } else {
            binding = TemplateTextSuggestionsBinding.bind(itemView)
        }

        val suggestion = items.getOrNull(position) ?: return Space(context)
        queryText?.also { query ->
            binding.run {
                title.text = getSpannable(query, suggestion.title)
                url.isVisible = !suggestion.url.isNullOrEmpty()
                url.text = getSpannable(query, suggestion.url ?: "")
                if (suggestion.icon == null) {
                    icon.setImageResource(R.drawable.search)
                    icon.setColorFilter(context.getOnSurfaceColor())
                } else {
                    icon.setImageBitmap(suggestion.icon)
                    icon.clearColorFilter()
                }

                autocompleteButton.isVisible = !url.isVisible
                autocompleteButton.setOnClickListener {
                    context.onUrlUpdated(suggestion.title, suggestion.title.length)
                }
            }
        }

        return itemView
    }

    private fun getSpannable(query: String, text: String): SpannableStringBuilder {
        val spannable = SpannableStringBuilder(text)
        val lcSuggestion = text.lowercase(Locale.getDefault())
        var queryTextPos = lcSuggestion.indexOf(query)
        while (queryTextPos >= 0) {
            spannable.setSpan(
                StyleSpan(Typeface.BOLD),
                queryTextPos,
                queryTextPos + query.length,
                Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
            )
            queryTextPos = lcSuggestion.indexOf(query, queryTextPos + query.length)
        }
        return spannable
    }

    override fun getFilter(): Filter {
        return filter
    }

    private inner class ItemFilter : Filter() {
        override fun performFiltering(constraint: CharSequence?): FilterResults {
            val filterResults = FilterResults()

            constraint?.takeUnless { it.isBlank() }?.let {
                val provider: SuggestionProvider = provider
                val query = constraint.toString().lowercase(Locale.getDefault()).trim { it <= ' ' }
                if (!enableFiltering || UrlUtils.isUriSupported(query)
                    || !context.settingsPreference.getIntBool(SettingsKeys.useSearchSuggestions))
                    return cancelFiltering()

                val results = provider.fetchResults(query)
                filterResults.count = items.size
                filterResults.values = items
                queryText = query
                items = results
            }
            return filterResults
        }

        fun cancelFiltering(): FilterResults {
            queryText = ""
            items = listOf()
            return FilterResults().apply {
                count = 0
                values = null
            }
        }

        override fun publishResults(constraint: CharSequence?, results: FilterResults) {
            notifyDataSetChanged()
        }
    }

    private val provider: SuggestionProvider
        get() = SuggestionProvider(context) { error, throwable ->
            MainScope().launch {
                Snackbar.make(
                    context.webviewContainer,
                    error, Snackbar.LENGTH_SHORT
                ).run {
                    setBehavior(BaseTransientBottomBar.Behavior().apply {
                        setSwipeDirection(SwipeDismissBehavior.SWIPE_DIRECTION_ANY)
                    })
                    setAction(R.string.customactivityoncrash_error_activity_error_details_title) {
                        context.copyClipboard(throwable.message)
                    }
                    setStartAligned()
                    show()
                }
            }
        }
}